ext/url/ada_url.cpp

```diff
@@ -0,0 +1,197 @@
+/*
+   +----------------------------------------------------------------------+
+   | Copyright (c) The PHP Group                                          |
+   +----------------------------------------------------------------------+
+   | This source file is subject to version 3.01 of the PHP license,      |
+   | that is bundled with this package in the file LICENSE, and is        |
+   | available through the world-wide-web at the following url:           |
+   | https://www.php.net/license/3_01.txt                                 |
+   | If you did not receive a copy of the PHP license and are unable to   |
+   | obtain it through the world-wide-web, please send a note to          |
+   | license@php.net so we can mail you a copy immediately.               |
+   +----------------------------------------------------------------------+
+   | Authors: Máté Kocsis <kocsismate@php.net>                            |
+   +----------------------------------------------------------------------+
+*/
+
+#ifdef HAVE_CONFIG_H
+# include "config.h"
+#endif
+
+#include <cstdlib>
+#include <string.h>
+#include <cctype>
+#include <sys/types.h>
+
+extern "C" {
+	#include "php.h"
+	#include "ada_url.h"
+	#include "Zend/zend_enum.h"
+	#include "Zend/zend_interfaces.h"
+	#include "url_arginfo.h"
+}
+
+#include <ada/ada.h>
+
+static zend_class_entry *url_ce;
+static zend_class_entry *url_component_ce;
+static zend_class_entry *url_parser_ce;
+
+void parseUrlComponentsArray(zend_string *url, zval *return_value)
+{
+	std::string_view protocol, hostname, username, password, port, path, query, fragment;
+
+	ada::result<ada::url_aggregator> ada_url = ada::parse<ada::url_aggregator>(ZSTR_VAL(url));
+	if (!ada_url) {
+		RETURN_NULL();
+	}
+
+	protocol = ada_url->get_protocol();
+	hostname = ada_url->get_hostname();
+	port = ada_url->get_port();
+	username = ada_url->get_username();
+	password = ada_url->get_password();
+	path = ada_url->get_protocol();
```

Comment:
isn't it a copy paste error ?

Reply:
Ah yes, good catch! I'll ignore it though since using ADA would cause a performance degradation :/